standalone/prb/src/tx.rs

```diff
@@ -0,0 +1,779 @@
+            *running_txs = current_txs.clone();
+
+            for _ in current_txs.iter() {
+                let _ = pending_txs.pop_front();
+            }
+            for i in last_running_txs {
+                past_txs.push_front(i);
+            }
+
+            drop(pending_txs);
+            drop(running_txs);
+            drop(past_txs);
+
+            let mut tx_map: StdHashMap<u64, Vec<usize>> = StdHashMap::new();
+            for i in current_txs {
+                let tx = self.tx_map.get(&i).ok_or(UnknownDataMismatch)?;
+                let tx = tx.lock().await;
+                let pid = tx.pid;
+                drop(tx);
+                if let Some(group) = tx_map.get_mut(&pid) {
+                    group.push(i);
+                } else {
+                    let group = vec![i];
+                    let _ = tx_map.insert(pid, group);
+                };
+            }
+            join_all(
+                tx_map
+                    .into_iter()
+                    .map(|(pid, v)| {
+                        let self_move = self.clone();
+                        async move {
+                            if let Err(e) = self_move.clone().wrap_send_tx_group(pid, v).await {
+                                error!("wrap_send_tx_group: {e}");
+                                std::process::exit(255);
+                            }
+                        }
+                    })
+                    .collect::<Vec<_>>(),
+            )
+            .await;
+
+            let running_txs = self.running_txs.clone();
+            let mut running_txs = running_txs.lock().await;
+            let past_txs = self.past_txs.clone();
+            let mut past_txs = past_txs.lock().await;
+
+            let last_running_txs = running_txs.clone();
+            *running_txs = Vec::new();
+            for i in last_running_txs {
+                past_txs.push_front(i);
+            }
+            drop(running_txs);
+            drop(past_txs);
+        }
+        error!("Unexpected exit of start_trader!");
+        std::process::exit(255);
+    }
+    async fn wrap_send_tx_group(self: Arc<Self>, pid: u64, ids: Vec<usize>) -> Result<()> {
+        if ids.is_empty() {
+            anyhow::bail!("TxGroup can't be empty!");
+        }
+
+        for id in ids.clone() {
+            let tx = self.tx_map.get(&id).ok_or(UnknownDataMismatch)?;
+            let mut tx = tx.lock().await;
+            tx.state = TransactionState::Running;
+            drop(tx);
+        }
+
+        match self.clone().send_tx_group(pid, ids.clone()).await {
+            Ok(ret) => {
+                for (idx, r) in ret.into_iter().enumerate() {
+                    let id = ids.get(idx).ok_or(UnknownDataMismatch)?;
+                    let tx = self.clone().tx_map.get(id).ok_or(UnknownDataMismatch)?;
+                    let mut tx = tx.lock().await;
+                    let shot = tx.shot.swap(None).ok_or(UnknownDataMismatch)?;
```

Comment:
Since the `tx` is already mutex-protected data, it might be unnecessary to wrap the `tx.shot` in an AtomicCell?